src/validators/union.rs

```diff
@@ -305,9 +338,17 @@ impl Validator for TaggedUnionValidator {
                         match lookup_key.$get_method($( $dict ),+)? {
                             Some((_, value)) => {
                                 if self.strict {
-                                    value.strict_str()
+                                    if let Ok(val_int) = value.strict_int() {
+                                        Ok(ChoiceKey::Int(val_int))
+                                    } else {
+                                        Ok(ChoiceKey::Str(value.strict_str()?.as_cow()?.as_ref().to_string()))
+                                    }
                                 } else {
-                                    value.lax_str()
+                                    if let Ok(val_int) = value.lax_int() {
+                                        Ok(ChoiceKey::Int(val_int))
+                                    } else {
+                                        Ok(ChoiceKey::Str(value.lax_str()?.as_cow()?.as_ref().to_string()))
+                                    }
                                 }
```

Comment:
probably easier here to remove the `if self.struct`, and instead do

```rs
if let Ok(int) = value.validate_int(self.strict) {
    Ok(ChoiceKey::Int(val_int))
} else {
    Ok(ChoiceKey::Str(value.validate_str(self.strict)?.as_cow()?.as_ref().to_string()))
}

Reply:
will do